### PR TITLE
Add a skeleton generic netlink decoder

### PR DESCRIPTION
**Pull request by donaldh**
Add a new netlink decoder for generic netlink messages with initial support for nlctrl message decoding and hook it into the table of netlink_decoders.

### CODE REVIEW

File: src/netlink.c [hunk 1]
```diff
@@ -596,7 +596,8 @@ static const netlink_decoder_t netlink_decoders[] = {
 	[NETLINK_NETFILTER] = decode_netlink_netfilter,
 	[NETLINK_ROUTE] = decode_netlink_route,
 	[NETLINK_SELINUX] = decode_netlink_selinux,
-	[NETLINK_SOCK_DIAG] = decode_netlink_sock_diag
+	[NETLINK_SOCK_DIAG] = decode_netlink_sock_diag,
+	[NETLINK_GENERIC] = decode_netlink_generic
```
Comment (masatake) [inline src/netlink.c:600]: The array is sorted by the name of index. So putting NETLINK_GENERIC next to NETLINK_CRYPTO.

File: src/defs.h [hunk 2]
```diff
@@ -1469,6 +1469,7 @@ DECL_NETLINK(netfilter);
 DECL_NETLINK(route);
 DECL_NETLINK(selinux);
 DECL_NETLINK(sock_diag);
+DECL_NETLINK(generic);
```
Comment (masatake) [inline src/defs.h:1472]: The name of the first argument sorts the sequence of DECL_NETLINK.
So putting `DECL_NETLINK(generic);` after `DECL_NETLINK(crypto);` may be better.

File: src/netlink_generic.c [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	tprint_struct_begin();
+	tprints_field_name("cmd");
+	printxval(NULL, cmd, "???");
+	tprint_struct_next();
+	tprints_field_name("version");
+	tprintf_string("%d", version);
+
+	if (len > 0) {
+		tprint_struct_next();
+		tprints_field_name("data");
```
Comment (masatake) [inline src/netlink_generic.c:36]: I had studied the way to decode the "data" area.
https://github.com/masatake/strace/commits/netlink-genl

Comment (masatake) [reply src/netlink_generic.c:36]: I found this is a fallback path. Please, ignore what I wrote. Sorry for making noise.

**Comment by masatake** [comment]
I tried this change with `$ strace genl ctrl list`. It worked fine.

File: src/netlink.c (hunk 1 above)
Comment (donaldh) [reply src/netlink.c:600]: Thanks, I will resin with a fix for this

File: src/defs.h (hunk 2 above)
Comment (donaldh) [reply src/defs.h:1472]: Yep, will fix this.

File: src/netlink_generic.c (hunk 3 above)
Comment (donaldh) [reply src/netlink_generic.c:36]: No worries, thanks for checking.

**Comment by donaldh** [comment]
I'm working on nlattr tests for this.

**Comment by masatake** [comment]
Could you look at https://github.com/donaldh/strace/pull/1 ?

**Comment by masatake** [comment]
Another implementation is submitted: https://lists.strace.io/pipermail/strace-devel/2024-June/011478.html

**Comment by masatake** [comment]
Wow!

**Comment by masatake** [comment]
I tried the new commits. When I run netlink_nlctrl.gen.test directly I got:
```
[yamato@dev64]~/var/strace/tests% ./netlink_nlctrl.gen.test 
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
[yamato@dev64]~/var/strace/tests% cat /etc/redhat-release 
Fedora release 40 (Forty)
```

File: src/netlink_generic.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+{
+	struct genlmsghdr h;
+
+	if (len < GENL_HDRLEN || nlmsghdr->nlmsg_type == NLMSG_DONE)
+		return false;
+
+	if (!umove_or_printaddr(tcp, addr, &h)) {
+		const char *family =
+			xlookup(genl_families_xlat(tcp), nlmsghdr->nlmsg_type);
+		netlink_generic_decoder_t decoder = lookup_genl_decoder(family);
```
Comment (masatake) [inline src/netlink_generic.c:68]: I think using nlmsg_type, an integer as a key is better than using family, a C string.
I would like to see:

https://github.com/masatake/strace/commit/c96f3c7d4c7257d7ac7803b9d249a0a1a2e10eac
https://github.com/masatake/strace/commit/2d2e06bbeef025c81d565378d6ee2bac52c2aa56

**Comment by ldv-alt** [comment]
> I tried the new commits. When I run netlink_nlctrl.gen.test directly I got:
> 
> ```
> [yamato@dev64]~/var/strace/tests% ./netlink_nlctrl.gen.test 
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d0ffd
> ../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
> ../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
> ../../src/strace: umoven: short read (1 < 2) @0x7fe4e16d6fff
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> ../../src/strace: umoven: short read (3 < 4) @0x7fe4e16d6ffd
> [yamato@dev64]~/var/strace/tests% cat /etc/redhat-release 
> Fedora release 40 (Forty)
> ```

This is expected.  In fact, many tests have some coverage for corner cases.

File: src/netlink_generic.c (hunk 4 above)
Comment (ldv-alt) [reply src/netlink_generic.c:68]: This approach can also be used in src/netlink_nlctrl.c to replace CTRL_ATTR_FAMILY_NAME with CTRL_ATTR_FAMILY_ID for family-dependent CTRL_ATTR_OP_ID decoding.

File: src/netlink_nlctrl.c [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+#include "xlat/genl_thermal_cmd.h"
+#include "xlat/nl_attr_type.h"
+#include "xlat/nl_policy_type_attr.h"
+
+#define ARG_PAIR(arg) { arg, arg }
+static const struct {
+	const char *family;
+	const struct xlat *cmd[2];
+} family_names[] = {
+	{ "nlctrl", ARG_PAIR(genl_ctrl_cmd) },
```
Comment (masatake) [inline src/netlink_nlctrl.c:47]: We can unify `decoders` in lookup_genl_decoder() in src/netlink_generic.c and `family_names`.

File: src/netlink_nlctrl.c [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	return decode_nla_str(tcp, addr, len, NULL);
+}
+
+static
+DECL_NLA(ctrl_attr_op_id)
+{
+	const unsigned int *idx = opaque_data;
+	struct decode_nla_xlat_opts opts = {
+		.xlat = family_names[*idx].cmd[!entering(tcp)],
+		.dflt = "CTRL_CMD_???",
```
Comment (masatake) [inline src/netlink_nlctrl.c:85]: It would be nice if we could define family-specific .dflt value in the `family_names` array.

```
static const struct {
	const char *family;
	const struct xlat *cmd[2];
        const char *dflt;
} family_names[] = {
...
	{ THERMAL_GENL_FAMILY_NAME, ARG_PAIR(genl_thermal_cmd), "THERMAL_GENL_CMD_???" },
...
}
...

	struct decode_nla_xlat_opts opts = {
		.xlat = family_names[*idx].cmd[!entering(tcp)],
		.dflt = family_names[*idx].dflt,
        ...
```

File: src/netlink_generic.c (hunk 4 above)
Comment (masatake) [reply src/netlink_generic.c]: YES!

File: src/netlink_nlctrl.c (hunk 5 above)
Comment (masatake) [reply src/netlink_nlctrl.c]: Here is my version: https://github.com/masatake/strace/blob/b82b7a73e8ebb2f16c0ae42424616e75ef5b3495/src/netlink_generic.h#L15 .

If the direction is acceptable, I will make a new pull request after you merge this pull request.

Comment (ldv-alt) [reply src/netlink_nlctrl.c]: The only complication I see is testing because strace will be able to resolve identifiers only for those genl families that are registered in the kernel.  How do we test decoding of other genl families?

**Review by ldv-alt — APPROVED** [review]
